### ISSUE TITLE
IntHashSet performance improvements

### ISSUE THREAD

**Issue by feribg** [real-logic/Agrona#84]
I'm not sure if this is within the scope of the project but the performance of some primitive collections can be dramatically improved if their scope is extended a bit outside of the interface they implement. 

The concrete case we had was intersection of a lot of sets, so mainly calling `addAll` and `removeAll`. These methods do allocate because `disjunction` is generic and hence calls `IntIterator.next()` which boxes the ints.

The following slightly improved version of addAll that now takes a concrete type and not a Collection, performs 2x+ as fast and produces no real garbage as compared to about 112Bytes / op in the current implementation.

```
 /**
     * {@inheritDoc}
     */
    public boolean addAll(BetterIntHashSet coll)
    {
        Objects.requireNonNull(coll);

        boolean acc = false;

        int t;
        for(IntIterator var3 = coll.iterator(); var3.hasNext(); acc |= add(t)) {
            t = var3.nextValue(); //key line
        }

        return acc;

    }
```

Some JMH benchmarks, where initial set is 20000 random integers, medium set is 500 random integers (half existing in the bigger set), small set is 50 (half existing):

```
Benchmark                                                        Mode  Cnt         Score          Error   Units
MyBenchmark.mediumBetterIntSet                                  thrpt    3  61737303.233 ± 25860243.729   ops/s
MyBenchmark.mediumBetterIntSet:·gc.alloc.rate                   thrpt    3         0.004 ±        0.001  MB/sec
MyBenchmark.mediumBetterIntSet:·gc.alloc.rate.norm              thrpt    3        ≈ 10⁻⁴                   B/op
MyBenchmark.mediumBetterIntSet:·gc.count                        thrpt    3           ≈ 0                 counts
MyBenchmark.mediumIntSet                                        thrpt    3  49261883.663 ±  6123491.825   ops/s
MyBenchmark.mediumIntSet:·gc.alloc.rate                         thrpt    3      5259.704 ±      649.795  MB/sec
MyBenchmark.mediumIntSet:·gc.alloc.rate.norm                    thrpt    3       112.000 ±        0.001    B/op
MyBenchmark.mediumIntSet:·gc.churn.PS_Eden_Space                thrpt    3      5295.205 ±     2190.657  MB/sec
MyBenchmark.mediumIntSet:·gc.churn.PS_Eden_Space.norm           thrpt    3       112.764 ±       53.327    B/op
MyBenchmark.mediumIntSet:·gc.churn.PS_Survivor_Space            thrpt    3         0.311 ±        1.962  MB/sec
MyBenchmark.mediumIntSet:·gc.churn.PS_Survivor_Space.norm       thrpt    3         0.007 ±        0.043    B/op
MyBenchmark.mediumIntSet:·gc.count                              thrpt    3        42.000                 counts
MyBenchmark.mediumIntSet:·gc.time                               thrpt    3        33.000                     ms
MyBenchmark.mediumObjSet                                        thrpt    3    688364.025 ±   350004.192   ops/s
MyBenchmark.mediumObjSet:·gc.alloc.rate                         thrpt    3         0.004 ±        0.001  MB/sec
MyBenchmark.mediumObjSet:·gc.alloc.rate.norm                    thrpt    3         0.006 ±        0.004    B/op
MyBenchmark.mediumObjSet:·gc.count                              thrpt    3           ≈ 0                 counts
MyBenchmark.removeBetterSmallIntSet                             thrpt    3  82975881.658 ±  4943074.441   ops/s
MyBenchmark.removeBetterSmallIntSet:·gc.alloc.rate              thrpt    3         0.004 ±        0.001  MB/sec
MyBenchmark.removeBetterSmallIntSet:·gc.alloc.rate.norm         thrpt    3        ≈ 10⁻⁴                   B/op
MyBenchmark.removeBetterSmallIntSet:·gc.count                   thrpt    3           ≈ 0                 counts
MyBenchmark.removeSmallIntSet                                   thrpt    3  42362666.236 ±  2309801.182   ops/s
MyBenchmark.removeSmallIntSet:·gc.alloc.rate                    thrpt    3      3876.827 ±      213.370  MB/sec
MyBenchmark.removeSmallIntSet:·gc.alloc.rate.norm               thrpt    3        96.000 ±        0.001    B/op
MyBenchmark.removeSmallIntSet:·gc.churn.PS_Eden_Space           thrpt    3      3902.694 ±     3730.180  MB/sec
MyBenchmark.removeSmallIntSet:·gc.churn.PS_Eden_Space.norm      thrpt    3        96.632 ±       87.453    B/op
MyBenchmark.removeSmallIntSet:·gc.churn.PS_Survivor_Space       thrpt    3         0.218 ±        0.568  MB/sec
MyBenchmark.removeSmallIntSet:·gc.churn.PS_Survivor_Space.norm  thrpt    3         0.005 ±        0.014    B/op
MyBenchmark.removeSmallIntSet:·gc.count                         thrpt    3        21.000                 counts
MyBenchmark.removeSmallIntSet:·gc.time                          thrpt    3        24.000                     ms
MyBenchmark.removeSmallObjSet                                   thrpt    3  10655251.346 ±   400175.869   ops/s
MyBenchmark.removeSmallObjSet:·gc.alloc.rate                    thrpt    3         0.004 ±        0.002  MB/sec
MyBenchmark.removeSmallObjSet:·gc.alloc.rate.norm               thrpt    3        ≈ 10⁻³                   B/op
MyBenchmark.removeSmallObjSet:·gc.count                         thrpt    3           ≈ 0                 counts
MyBenchmark.smallBetterIntSet                                   thrpt    3  73270433.394 ±  9965782.861   ops/s
MyBenchmark.smallBetterIntSet:·gc.alloc.rate                    thrpt    3         0.004 ±        0.001  MB/sec
MyBenchmark.smallBetterIntSet:·gc.alloc.rate.norm               thrpt    3        ≈ 10⁻⁴                   B/op
MyBenchmark.smallBetterIntSet:·gc.count                         thrpt    3           ≈ 0                 counts
MyBenchmark.smallIntSet                                         thrpt    3  36804813.855 ± 13066320.200   ops/s
MyBenchmark.smallIntSet:·gc.alloc.rate                          thrpt    3      3929.380 ±     1397.023  MB/sec
MyBenchmark.smallIntSet:·gc.alloc.rate.norm                     thrpt    3       112.000 ±        0.001    B/op
MyBenchmark.smallIntSet:·gc.churn.PS_Eden_Space                 thrpt    3      4030.791 ±     6459.226  MB/sec
MyBenchmark.smallIntSet:·gc.churn.PS_Eden_Space.norm            thrpt    3       114.790 ±      143.200    B/op
MyBenchmark.smallIntSet:·gc.churn.PS_Survivor_Space             thrpt    3         0.156 ±        1.507  MB/sec
MyBenchmark.smallIntSet:·gc.churn.PS_Survivor_Space.norm        thrpt    3         0.004 ±        0.042    B/op
MyBenchmark.smallIntSet:·gc.count                               thrpt    3        24.000                 counts
MyBenchmark.smallIntSet:·gc.time                                thrpt    3        27.000                     ms
MyBenchmark.smallObjSet                                         thrpt    3   5860549.692 ±  2416859.155   ops/s
MyBenchmark.smallObjSet:·gc.alloc.rate                          thrpt    3         0.004 ±        0.001  MB/sec
MyBenchmark.smallObjSet:·gc.alloc.rate.norm                     thrpt    3         0.001 ±        0.001    B/op
MyBenchmark.smallObjSet:·gc.count                               thrpt    3           ≈ 0                 counts
```

**Comment by mjpt777** [comment]
We would be happy to accept a pull request for this type of extension.

**Comment by feribg** [comment]
@mjpt777 
Thanks for cleaning up the PR, I saw the changes you made, actually something is weird in JMH but I haven't had time to delve deeper, maybe you have a clue. I originally had it the way you rewrote them, directly looping over the underlying array instead of using iterator. For some reason JMH reports the iterator version to be much faster, which doesn't make any sense:

```        Objects.requireNonNull(other);
        boolean acc = false;
        final int missingValue = other.missingValue;
        for (final int value : other.values)
        {
            if(value != missingValue)
                acc |= add(value);
        }

        return acc;
```
below:
```
Benchmark                                 Mode  Cnt        Score         Error  Units
MyBenchmark.mediumBetterIntSet           thrpt    3   943859.818 ±  901538.380  ops/s
MyBenchmark.mediumBetterIntSet:JFR       thrpt               NaN                  N/A
MyBenchmark.removeBetterSmallIntSet      thrpt    3  9722493.660 ± 4274660.673  ops/s
MyBenchmark.removeBetterSmallIntSet:JFR  thrpt               NaN                  N/A
MyBenchmark.smallBetterIntSet            thrpt    3  9603668.164 ± 3397169.451  ops/s
MyBenchmark.smallBetterIntSet:JFR        thrpt               NaN                  N/A
```

```
Objects.requireNonNull(other);

        boolean acc = false;

        int t;
        for(IntIterator var3 = other.iterator(); var3.hasNext(); acc |= add(t)) {
            t = var3.nextValue();
        }

        return acc;
```
below:
```
Benchmark                                 Mode  Cnt         Score          Error  Units
MyBenchmark.mediumBetterIntSet           thrpt    3  66048651.300 ± 33846633.694  ops/s
MyBenchmark.mediumBetterIntSet:JFR       thrpt                NaN                   N/A
MyBenchmark.removeBetterSmallIntSet      thrpt    3  81282630.096 ± 25498509.279  ops/s
MyBenchmark.removeBetterSmallIntSet:JFR  thrpt                NaN                   N/A
MyBenchmark.smallBetterIntSet            thrpt    3  62508465.421 ± 54780837.005  ops/s
MyBenchmark.smallBetterIntSet:JFR        thrpt                NaN                   N/A
```

I have attached some flight recordings of the JMH. Looping over the array should be much faster, so Im guessing there is an issue the way im doing the tests.
[Archive.zip](https://github.com/real-logic/Agrona/files/661854/Archive.zip)

**Comment by mjpt777** [comment]
Can you post a gist of your benchmark?

**Comment by feribg** [comment]
https://gist.github.com/feribg/26957e6c5e70c9723d152176a2969532

**Comment by mjpt777** [comment]
After the first iteration the test will always be adding/removing the same data that is already in the set so nothing more will be added/remove there the set is not being modified. I don't think these tests are really representative.  It would take more extensive testing to get a realistic view of the differences.

**Comment by feribg** [comment]
Thanks, changed it to regenerate the state per invocation and now the results make a lot more sense. The direct loop over the array is orders of magnitude better with roughly fixed cost.

**Comment by mjpt777** [comment]
BTW this was released in 0.9.1 yesterday.
